cdap-data-fabric/src/main/java/io/cdap/cdap/spi/metadata/dataset/SearchHelper.java

```diff
@@ -304,22 +313,62 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, SearchRequest r
   }
 
-  private Set<MetadataEntity> getSortedEntities(List<MetadataEntry> results, SortInfo sortInfo) {
+  private Map<MetadataEntity, Map<String, Integer>> hashResults(List<MetadataResultEntry> results) {
+    Map<MetadataEntity, Map<String, Integer>> hashedResults = new LinkedHashMap<>();
+    for (MetadataResultEntry resultEntry : results) {
+      MetadataEntity entity = resultEntry.getMetadataEntity();
+      String searchTerm = resultEntry.getSearchTerm();
+      Map<String, Integer> map = hashedResults.getOrDefault(entity, new LinkedHashMap<>());
+      map.put(searchTerm, map.getOrDefault(searchTerm, 0) + 1);
+      hashedResults.putIfAbsent(entity, map);
+    }
+
+    return hashedResults;
+  }
+
+  /***
+   * Remove entities that do not have the required metadata tags
+   *
+   * @param results search results
+   * @param request users original search request
+   * @return map of results to query terms, without results that do not satisfy all query requirements
+   */
+  private Map<MetadataEntity, Map<String, Integer>> filterEntries(Map<MetadataEntity, Map<String, Integer>> results,
+                                                                  SearchRequest request) {
+    List<QueryTerm> queryTerms = QueryParser.parse(request.getQuery());
+    Set<String> requiredTerms = queryTerms.stream()
+      .filter(queryTerm -> queryTerm.getQualifier() == QueryTerm.Qualifier.REQUIRED)
+      .map(QueryTerm::getTerm).collect(Collectors.toSet());
+
+    Map<MetadataEntity, Map<String, Integer>> filteredResults = new LinkedHashMap<>();
+    for (Map.Entry<MetadataEntity, Map<String, Integer>> entity : results.entrySet()) {
+      if (entity.getValue().keySet().containsAll(requiredTerms)) {
+        filteredResults.put(entity.getKey(), entity.getValue());
+      }
+    }
+
+    return filteredResults;
+  }
+
+  private Set<MetadataEntity> getSortedEntities(Map<MetadataEntity, Map<String, Integer>> results, SortInfo sortInfo) {
     // if sort order is not weighted, return entities in the order received.
     // in this case, the backing storage is expected to return results in the expected order.
     if (SortInfo.SortOrder.WEIGHTED != sortInfo.getSortOrder()) {
       Set<MetadataEntity> entities = new LinkedHashSet<>(results.size());
-      for (MetadataEntry metadataEntry : results) {
-        entities.add(metadataEntry.getMetadataEntity());
+      for (MetadataEntity entity : results.keySet()) {
+        entities.add(entity);
       }
       return entities;
     }
     // if sort order is weighted, score results by weight, and return in descending order of weights
     // Score results
     final Map<MetadataEntity, Integer> weightedResults = new HashMap<>();
-    for (MetadataEntry metadataEntry : results) {
-      weightedResults.put(metadataEntry.getMetadataEntity(),
-                          weightedResults.getOrDefault(metadataEntry.getMetadataEntity(), 0) + 1);
+    for (MetadataEntity entity : results.keySet()) {
```

Comment:
Instead of looping over `keySet` and then call `get` by the key, you should loop over the `Map.entrySet` directly to get access to both key and value at the same time.

Reply:
got it, thank you!